---
output:
  word_document: default
editor_options: 
  chunk_output_type: console
---
```{r, include=FALSE}
#1. Load required libraries
#Old version of vegan
#require(devtools)
#install_version("vegan", version ="2.4-5", repos = "http://cran.us.r-project.org")

library(tm)
library(vegan)
library("biom")
library(tidyr)
library(ggplot2)
library(cowplot)
library(scales)
library(phyloseq)
library(backports)
library(qiime2R)
library(tibble)

#Source MiSeqR
#http://deneflab.github.io/MicrobeMiseq/
source("sources/miseqR.R")
load("workspace.RData")
```
QC Report


#Project Information
ProjectID: NP0440-MB4

Source: sTVOM

Date Reported:

Pipeline: Version2

QIIME: 2017.11

#Metadata
##Summary
###Sample Type
Sample types for the project included study samples, robogut, artifical colony, extraction blank, sequencing blank, and non-template controls.

###Before filtering
Prior to filtering, these were the number of samples observed:

###After Filtering
After filtering to a minimum of 10,000 reads, only taxa with a mean absolute abundance greater than 0.1% were included. Taxonomically, only taxa in the Bacteria kingdom, that were not in the chrloroplast class or mitochondrial family were included. Rarefying to 20,000 was implemented, to more clearly cluster samples according to biological origin, by drawing without replacement from each sample so that all samples have the same total count. The maximum rarefaction rate was set to 54,000 (the highest count of all samples) for this particular dataset. A citation on measurements of richness can be found here: https://onlinelibrary.wiley.com/doi/full/10.1046/j.1461-0248.2001.00230.x

After filtering, these were the number of samples observed:

###Variables
The filtered cohort was broken down by the following variables:


#Results
##Sequencing Depth
This image shows the sequencing depth prior to filtering:
![](Graphs/SeqDepth_prefilter.tiff)

This image shows the depth after filtering:
![](Graphs/SeqDepth_filter.tiff)

##Alpha Diversity
###By Sample Types
This image shows alpha diversity metrics by sample type:
![](Graphs/sampletype_alphadiv.tiff)

Viewing the dataset by extraction and sequencing batches, helps to ensure consistency between processing:

###By Extraction Batch:
![](Graphs/alpha_extractionbatch.tiff)

###By Sequencing Batch:
![](Graphs/alpha_sequencingbatch.tiff)


###Top OTUs by Genus
The top 30 OTUs across all study samples, in terms of abundance, were graphed by Genus:
![](Graphs/study_topgenus.tiff)


##Beta Diversity - pCOA Graphs
Beta diversity was measured using both quantitative (weighted UniFrac and Bray-Curtis) and qualitative measures (unweighted UniFrac and Binomial).Quantitative measures are ideally suited to revealing community differences that are due to changes in relative taxon abundance (e.g., when a particular set of taxa flourish because a limiting nutrient source becomes abundant). Quantitative measures are shown below.

###Bray Curtis
Bray-Curtis dissimilarity creates a matrix of the differences in microbial abundances between two samples. A 0 indicates that the samples share the same species at the same abundances, whereas a 1 indicates that both samples have complete different speices abundances.

####Results by Sample Type:

![](Graphs/sampletype_bray.tiff)

####Results by subject (Replicates Only):

![](Graphs/reps_bray.tiff)


###Weighted Unifrac
The UniFrac measures the phylogenetic distance between sets of taxa in a phylogenetic tree as the fraction of the branch length of the tree that leads to descendants from either one environment.It differs from dissimilarity measures such as Bray-Curtis, in that it incorporates information on the relative relatedness of community members by incorporating phylogenetic distances between observed organisms in the computation.Finally, this weighted version of the UniFrac metric accounts for the relative abundance of each of the taxa within the communities.

####Results by Sample Type:

![](Graphs/sampletype_weightunifrac.tiff)


####Results by subject (Replicates Only):

![](Graphs/reps_weightunifrac.tiff)


#Additional Information
Viewing the dataset by other feature variables, provided the following insights:
![](Graphs/alpha_collection.tiff)
![](Graphs/alpha_instrument.tiff)
![](Graphs/alpha_month.tiff)
![](Graphs/summary_genus_-20C.tiff)
![](Graphs/summary_genus_-80C.tiff)




