Workflow summary
================

1. Manifest management:

    * Manifest provided in config.yaml is checked for compliance with QIIME2 specifications
    * Per-flow cell manifests are created

2. Symlink fastqs to be viewable by DCEG PIs
3. Import and demultiplex fastqs
4. For external data only, fix any unmatched paired reads (can be generated by fastq QC process)
5. Denoise
6. Merge feature and sequence tables across flow cells; drop samples with zero reads
7. Build multiple sequence alignment, then build rooted and unrooted phylogenetic trees
8. Perform alpha- and beta-diversity analysis, rarefaction, and taxonomic classification

